File: kotlin/prototype/src/UnderlinePen.kt
package prototype

class UnderlinePen(private val ulChar: Char) : Product() {
    override fun use(str: String) {
        val length: Int = str.toByteArray().size
        println("*$str*")
        print(" ")
        for (i in 0..(length - 1)) {
            print(ulChar)
        }
        println()
    }

    override fun createClone(): Product {
        try {
            return clone() as Product
        } catch (e : CloneNotSupportedException) {
            throw e
        }
    }
}

File: typescript/prototype/underline-pen.ts
import { Product } from './product';
import lodash from 'lodash';

export class UnderlinePen implements Product {
  constructor(private ulChar: string) {}

  use(str: string): void {
    const length: number = encodeURIComponent(str).replace(/%../g,"x").length;
    console.log(`*${str}*`);
    process.stdout.write(' ');
    for(let i = 0; i < length; i++) {
      process.stdout.write(this.ulChar);
    }
    console.log('');
  }

  createClone(): Product {
    return lodash.cloneDeep(this);
  }
}
